### PR TITLE
Add Thanos Store memcached mixin

### PR DESCRIPTION
**Pull request by kakkoyun**
This PR adds `withMemcachedIndexCache` to Thanos Store to use `memcached` as index cache.
Enables posting compression.

* [ ] I added CHANGELOG entry for this change.
* [ ] Change is not relevant to the end user.

## Changes

* Adds `withMemcachedIndexCache` mixin.
* Enables posting compression.
* Added flags from Thanos Store:
  * `--experimental.enable-index-cache-postings-compression`
  * `--index-cache.config=${CONFIG}'`

## Verification

* `make generate`

### CODE REVIEW

**Comment by brancz** [comment]
DCO missing, otherwise looks really awesome!

File: all.jsonnet [hunk 1]
```diff
@@ -7,7 +7,7 @@ local commonConfig = {
   config+:: {
     local cfg = self,
     namespace: 'thanos',
-    version: 'v0.10.1',
+    version: 'v0.12.0-rc.0',
```
Comment (lilic) [inline all.jsonnet:10]: Curious what happened with 0.11 release? :)

Comment (bwplotka) [reply all.jsonnet:10]: 0.11 is in quarantine. :mask:

Comment (kakkoyun) [reply all.jsonnet:10]: I will address this before merging this. Thanks @lilic.

File: all.jsonnet [hunk 2]
```diff
@@ -62,6 +62,16 @@ local s = t.store + t.store.withVolumeClaimTemplate + t.store.withServiceMonitor
   },
 };
 
+local swm = t.store + t.store.withVolumeClaimTemplate + t.store.withServiceMonitor + t.store.withMemcachedIndexCache + commonConfig + {
```
Comment (metalmatze) [inline all.jsonnet:65]: Let's add a few line breaks for readability.
```suggestion
local swm = 
  t.store + 
  t.store.withVolumeClaimTemplate + 
  t.store.withServiceMonitor + 
  t.store.withMemcachedIndexCache + 
  commonConfig + 
  {
```

File: jsonnet/kube-thanos/kube-thanos-store.libsonnet [hunk 3]
```diff
@@ -120,6 +120,56 @@ local k = import 'ksonnet/ksonnet.beta.4/k.libsonnet';
     },
   },
 
+  withMemcachedIndexCache:: {
+    local ts = self,
+    config+:: {
+      memcached+: {
+        addresses: error 'must provide memcached addresses',
+        timeout: '0s',
```
Comment (kakkoyun) [inline jsonnet/kube-thanos/kube-thanos-store.libsonnet:128]: @pracucci  @pstibrany I'm trying to come up with sane default for memcached, could point me to the right direction? What works for you best?

Comment (pracucci) [reply jsonnet/kube-thanos/kube-thanos-store.libsonnet:128]: It really depends on the scale you wanna address. I still think the defaults I've set into Thanos are OK for a small/medium scale:
https://github.com/thanos-io/thanos/blob/1a2edf19475b854181e5abaa4b6f1c9caf2a1547/pkg/cacheutil/memcached_client.go#L35-L44

If you wanna be a bit more aggressive you may consider:
```
timeout: '2s',
maxIdleConnections: 1000,
maxAsyncConcurrency: 100,
maxAsyncBufferSize: 100000,
maxGetMultiConcurrency: 900,
maxGetMultiBatchSize: 1000,
```

I've also seen you set `dnsProviderUpdateInterval: '3s'`. I'm not sure there's a real value of being that aggressive here. The default 10s could be enough (but maybe you find a good reason to lower it).

Comment (kakkoyun) [reply jsonnet/kube-thanos/kube-thanos-store.libsonnet]: Thanks a lot for the response, this was really helpful.

**Comment by kakkoyun** [comment]
This is in decent shape to merge, could you have another look @metalmatze?

File: all.jsonnet [hunk 4]
```diff
@@ -33,36 +33,60 @@ local b = t.bucket + commonConfig + {
...
+          t.compact.withServiceMonitor +
+          commonConfig + {
   config+:: {
     name: 'thanos-compact',
     replicas: 1,
   },
 };
 
-local re = t.receive + t.receive.withVolumeClaimTemplate + t.receive.withServiceMonitor + commonConfig + {
+local re = t.receive + t.receive.withVolumeClaimTemplate +
```
Comment (bwplotka) [inline all.jsonnet:46]: inconsistent new line with compact

Comment (bwplotka) [reply all.jsonnet:46]: or rather compact is inconsistent

File: all.jsonnet [hunk 5]
```diff
@@ -33,36 +33,60 @@ local b = t.bucket + commonConfig + {
...
-local q = t.query + t.query.withServiceMonitor + commonConfig + {
+local swm = t.store +
+            t.store.withVolumeClaimTemplate +
+            t.store.withServiceMonitor + t.store.withMemcachedIndexCache +
+            commonConfig + {
+  config+:: {
+    name: 'thanos-store',
+    replicas: 1,
+    memcached+: {
+      addresses: [],
```
Comment (bwplotka) [inline all.jsonnet:82]: can at least add comment how to add your addresses?

Or can we add some k8s memcached service ? Example one or maybe it should be part of mixin (memcached statefulset)

**Pull request merged**